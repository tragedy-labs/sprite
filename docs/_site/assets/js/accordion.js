/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
