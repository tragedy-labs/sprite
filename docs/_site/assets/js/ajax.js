// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
