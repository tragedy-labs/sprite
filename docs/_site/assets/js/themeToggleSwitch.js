function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
