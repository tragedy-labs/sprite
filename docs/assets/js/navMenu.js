document.addEventListener("DOMContentLoaded", function () {
  const content = document.getElementById("content");
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const links = mainNav.querySelectorAll("a");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = null;
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;

    const mainNavWidth = mainNav.offsetWidth;
    const ratio = (mainNavWidth + currentTranslation) / mainNavWidth;

    content.style.opacity = 0.65 * (1 - ratio) + 0.35;
  }

  function escHandler(e) {
    if (e.key === "Escape") {
      closeMenu();
    }
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
    content.addEventListener("touchstart", closeMenu);
    content.addEventListener("mousedown", closeMenu);
    document.addEventListener("keydown", escHandler);
    // TODO: This could be handled in the ajax script
    // since it already handles link clicks
    links.forEach((link) => {
      link.addEventListener("click", closeMenu);
    });
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
    content.removeEventListener("touchstart", closeMenu);
    content.removeEventListener("mousedown", closeMenu);
    document.removeEventListener("keydown", escHandler);
    links.forEach((link) => {
      link.removeEventListener("click", closeMenu);
    });
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
    content.style.opacity = null;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    content.classList.toggle("fade");
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
