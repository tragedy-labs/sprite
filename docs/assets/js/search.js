import data from "/assets/js/searchData.js";
import serializedIndex from "./assets/js/serialized.js";
//import lunr from "lunr"; // assuming lunr is a separate module

// Constants
const MIN_SEARCH_LENGTH = 3;

// State definition
const state = {
  filters: {
    class: true,
    method: true,
    type: true,
  },
  results: {
    filtered: [],
    full: [],
  },
  visible: false,
};

// Elements for reference
const elements = {
  input: null,
  resultsBox: null,
  resultList: null,
  numberOfResults: null,
  numberResultsShowing: null,
  backdrop: null,
  filters: null,
  searchLabel: null,
  close: null,
};

// Initialize search
async function initializeSearch() {
  const root = document.getElementById("search_button");
  elements.input = document.getElementById("searchInput");
  elements.resultsBox = document.getElementById("searchResults");
  elements.resultList = document.getElementById("resultList");
  elements.numberOfResults = document.getElementById("numberOfResults");
  elements.numberResultsShowing = document.getElementById("numberResultsShowing");
  elements.filters = document.getElementById("searchFilters")
    .getElementsByTagName("input");
  elements.searchLabel = document.getElementById("searchLabel");
  elements.close = elements.resultsBox.querySelector("#close_search_dialog");

  // Add event listeners
  elements.input.addEventListener("input", search);
  elements.close.addEventListener("click", hideSearchBox);

  const openSearch = document.getElementById('search_button');

  openSearch.addEventListener('click', showResultsBox);
  document.addEventListener('keydown', (e)=>{
    if (e.key === '/') {
      e.preventDefault();
      showResultsBox();
    };
  }, false);

  // Load the index
  const index = lunr.Index.load(serializedIndex);

  // Initialize search filter toggles
  for (let i = 0; i < elements.filters.length; ++i) {
    elements.filters[i].addEventListener("change", () => {
      state.filters[elements.filters[i].value] = elements.filters[i].checked;
      updateResultList();
    });
  }

  // Define functions
  function appendBackdrop() {
    elements.backdrop = document.createElement("div");
    elements.backdrop.id = "backdrop";
    elements.backdrop.addEventListener("click", hideSearchBox);
    document.body.appendChild(elements.backdrop);
  }

  function removeBackdrop() {
    elements.backdrop.removeEventListener("click", appendBackdrop);
    elements.backdrop.remove();
    elements.backdrop = undefined;
  }

  function handleKeyboardSelection() {
    const items = elements.list.querySelectorAll("a");
    if (items.length > 0) {
      if (document.activeElement === elements.input) {
        e.preventDefault();
        items[0].focus();
      } else {
        for (let i = 0; i < results.length; i++) {
          if (items[i] === document.activeElement) {
            e.preventDefault();
            if (i + 1 < items.length) {
              items[i + 1].focus();
            } else {
              elements.input.focus();
            }
            break;
          }
        }
      }
    }
  }

  function checkKeyPress(e) {
    if (e.key === "Tab") {
      handleKeyboardSelection(e);
    }
    if (e.key === "Escape") {
      hideSearchBox();
    }
  }

  function showResultsBox() {
    appendBackdrop();
    state.visible = true;
    elements.resultsBox.classList.remove("hidden");
    elements.input.focus();
    document.addEventListener("keydown", checkKeyPress);
  }

  function hideSearchBox() {
    state.visible = false;
    elements.resultsBox.classList.add("hidden");
    document.removeEventListener("keydown", checkKeyPress);
    removeBackdrop();
  }

  function updateResultList() {
    state.results.filtered = [];
    elements.resultList.innerHTML = "";
    // elements.numberOfResults.innerHTML = state.results.full.length;
    // elements.numberResultsShowing.innerHTML = state.results.filtered.length;

    if (state.results.full.length > 0) {
      state.results.full.forEach((result) => {
        const doc = data.find((doc) => doc.name === result.ref);
        const type = doc.type === "interface" ? "type" : doc.type;
        if (state.filters[type]) {
          state.results.filtered.push(result.ref);
          const section = document.createElement("section");
          const desc = document.createElement("p");
          const link = document.createElement("a");
          const li = document.createElement("li");
          li.classList.add(doc.type);

          link.href = `/sprite/doc/${doc.name}`;
          link.className = "searchResultLink";
          link.addEventListener("click", () => {});
          link.textContent = doc.parent
            ? `${doc.parent}.${doc.name}`
            : doc.name;

          desc.textContent = doc.desc;

          section.appendChild(link);
          section.appendChild(desc);

          li.appendChild(section);

          elements.resultList.appendChild(li);
        }
      });

      //elements.numberResultsShowing.innerHTML = state.results.filtered.length;
    }
  }

  function displayResults() {
    if (!state.visible) {
      showResultsBox();
    }
    updateResultList();
  }

  function search() {
    if (elements.input.value.length > 0) {
      elements.searchLabel.classList.add("hidden");
    } else {
      elements.searchLabel.classList.remove("hidden");
    }

    if (elements.input.value.length >= MIN_SEARCH_LENGTH) {
      const initial = index.search(elements.input.value);

      if (initial.length === 0) {
        const wildcardResults = index.search(`${elements.input.value}*`);
        state.results.full =
          wildcardResults.length > 0
            ? wildcardResults
            : index.search(`*${elements.input.value}*`);
      } else {
        state.results.full = initial;
      }

      displayResults();
    }
  }
}

document.addEventListener("DOMContentLoaded", initializeSearch);