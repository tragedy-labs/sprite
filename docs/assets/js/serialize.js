const data = require("./searchData.js");
const lunr = require("./lunr.js");

const fs = require("fs");
const path = require("path");

const config = {
  minSearchLength: 3,
  indexedFields: ["name", "desc", "type", "parent"],
};

// Build the index using Lunr.js
const index = lunr(function () {
  this.ref("name");

  config.indexedFields.forEach((field) => {
    this.field(field);
  }, this);

  data.forEach((doc) => {
    this.add(doc);
  }, this);
});

fs.writeFileSync(
  path.join(__dirname, "./serialized.json"),
  JSON.stringify(index, null, 2)
);
