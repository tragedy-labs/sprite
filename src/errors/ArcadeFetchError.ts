export interface ArcadeFetchErrorDescription {
  statusText: string;
  extended: string;
}

/**
 * The error thrown when the fetch() method of the Sprite class encounters an error.
 * @class
 * @param description A detailed error message to display.
 * @param response The response from the fetch() method.
 * @extends Error
 * @example
 * const response = await sprite.fetch();
 * if (response.status !== 200 || response.status !== 204) {
 *   throw new ArcadeFetchError("It wasn't a 204 or 200", response);
 * }
 */
const details: Record<number, ArcadeFetchErrorDescription> = {
  400: {
    statusText: 'Bad Request',
    extended: `The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing). In the context of ArcadeDB, this could be the result of the server being sent an invalid language, or an invalid command.`
  },
  403: {
    statusText: 'Forbidden / Invalid Credentials',
    extended: `The server understood the request but refuses to authorize it. If authentication credentials were provided in the request, the server considers them insufficient to grant access.`
  },
  404: {
    statusText: 'Not Found',
    extended: `The origin server did not find a current representation for the target resource or is not willing to disclose that one exists.`
  },
  500: {
    statusText: 'Internal Server Error',
    extended: `The server encountered an unexpected condition that prevented it from fulfilling the request.`
  }
};

/**
 * The error thrown when the fetch() method of the Sprite class encounters an error.
 */
export class ArcadeFetchError extends Error {
  detail: string;
  error: string;
  exception: string;
  constructor(response: Response) {
    const detail = details[response.status];
    const error = `${response.status} ${
      response.statusText ? response.statusText : detail.statusText
    }`;
    const errorString = `${error}. ${detail.extended}`;
    super(errorString);
    this.error = error;
    this.name = 'ArcadeFetchError';
    this.detail = detail.extended;
    this.exception = `(http) ${this.error}. https://httpstatuses.io/${response.status}`;
  }
}
