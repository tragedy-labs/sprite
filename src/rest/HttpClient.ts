import { ArcadeFetchError } from '../errors/ArcadeFetchError.js';
import { SpriteHeadersInit } from './SpriteHeaders.js';

export interface DatabaseSession {
  address: string;
  auth: string;
  databaseName: string;
  headers: HeadersInit;
}

export interface SpriteRequestInit extends RequestInit {
  headers: SpriteHeadersInit;
}

/*
 * NOTE:
 * It seems a bit redundant to have this class, but it's a good way to encapsulate
 * the HTTP request logic in one place. Additionally, this likely a stub for a
 * more complex implementation.
 */

/**
 * Utility class for making HTTP requests to the ArcadeDB server.
 */
class HttpClient {
  /**
   * Make a request to the server.
   * @param session The `SpriteDatabaseSession` to target in the request.
   * @param relativeEndpoint The relative {@link https://docs.arcadedb.com/#HTTP-API endpoint} for the request, if it's a database endpoint it includes the database name. Any composition must be handled by the calling method.
   * @param requestInit A `fetch` `RequestInit` object: `{ methods, headers, body }`.
   * @returns
   */
  public static async request(
    endpoint: string,
    requestInit: SpriteRequestInit
  ): Promise<Response> {
    try {
      const response: Response = await fetch(endpoint, requestInit);
      return this.handleResponse(response);
    } catch (error) {
      if (error instanceof ArcadeFetchError) {
        throw error;
      } else {
        throw new Error(
          'There was a problem fetching from the supplied database server. Check the address and try again.'
        );
      }
    }
  }
  /**
   * Handles error handling of the response, or returns the response if nothing went wrong.
   * @param response
   * @returns
   * @throws If there is a {@link https://httpstatuses.io/403 `403`} or {@link https://httpstatuses.io/404 `404`} error.
   * @see {@link ArcadeFetchError}
   */
  private static handleResponse(response: Response) {
    switch (response.status) {
      case 403:
      case 404:
        throw new ArcadeFetchError(response);
      default:
        return response;
    }
  }
}

export { HttpClient };
