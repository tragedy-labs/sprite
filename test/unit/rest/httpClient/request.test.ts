import { HttpClient, SpriteRequestInit } from '@/rest/HttpClient.js';
import { ArcadeFetchError } from '@/errors/ArcadeFetchError.js';

describe('HttpClient.request', () => {
  const endpoint = 'http://localhost:2480/api/vi/ready';

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should handle a POST request', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 200,
      json: async () => ({})
    } as Response);

    const init: SpriteRequestInit = {
      body: 'body',
      method: 'POST',
      headers: {
        Authorization: '',
        'Content-Type': 'application/json'
      }
    };

    await HttpClient.request(endpoint, init);

    expect(global.fetch).toHaveBeenCalledWith(endpoint, init);
  });

  it('should handle a GET request', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 200,
      json: async () => ({})
    } as Response);

    const init: SpriteRequestInit = {
      method: 'GET',
      headers: {
        Authorization: '',
        'Content-Type': 'application/json'
      }
    };

    await HttpClient.request(endpoint, init);
    expect(global.fetch).toHaveBeenCalledWith(endpoint, init);
  });

  it('should handle a 403 error response', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 403
    } as Response);

    const init: SpriteRequestInit = {
      body: 'body',
      method: 'POST',
      headers: {
        Authorization: '',
        'Content-Type': 'application/json'
      }
    };

    await expect(HttpClient.request(endpoint, init)).rejects.toThrow(
      ArcadeFetchError
    );
  });

  it('should handle a 404 error response', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 404
    } as Response);

    const init: SpriteRequestInit = {
      body: 'body',
      method: 'POST',
      headers: {
        Authorization: '',
        'Content-Type': 'application/json'
      }
    };

    await expect(HttpClient.request(endpoint, init)).rejects.toThrow(
      ArcadeFetchError
    );
  });
});
