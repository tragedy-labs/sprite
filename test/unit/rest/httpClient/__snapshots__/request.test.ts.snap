// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpClient.request should handle a 404 error response 1`] = `[Error: There was a problem fetching from the supplied database server. Check the address and try again.]`;
